libraries/standard/mqtt/src/iot_mqtt_api.c

```diff
@@ -1054,126 +1035,124 @@ IotMqttError_t IotMqtt_Connect( const IotMqttNetworkInfo_t * pNetworkInfo,
+        }
+    }
 
-    /* Add previous session subscriptions. */
-    if( pConnectInfo->pPreviousSubscriptions != NULL )
+    if( status == IOT_MQTT_SUCCESS )
     {
-        /* Previous subscription count should have been validated as nonzero. */
-        IotMqtt_Assert( pConnectInfo->previousSubscriptionCount > 0 );
+        /* Ensure the members set by operation creation and serialization
+         * are appropriate for a blocking CONNECT. */
+        IotMqtt_Assert( pOperation->u.operation.status == IOT_MQTT_STATUS_PENDING );
+        IotMqtt_Assert( ( pOperation->u.operation.flags & IOT_MQTT_FLAG_WAITABLE )
+                        == IOT_MQTT_FLAG_WAITABLE );
+        IotMqtt_Assert( pOperation->u.operation.periodic.retry.limit == 0 );
 
-        status = _IotMqtt_AddSubscriptions( pNewMqttConnection,
-                                            2,
-                                            pConnectInfo->pPreviousSubscriptions,
-                                            pConnectInfo->previousSubscriptionCount );
+        /* Set the operation type. */
+        pOperation->u.operation.type = IOT_MQTT_CONNECT;
 
-        if( status != IOT_MQTT_SUCCESS )
+        /* Add previous session subscriptions. */
+        if( pConnectInfo->pPreviousSubscriptions != NULL )
         {
-            goto cleanup;
+            /* Previous subscription count should have been validated as nonzero. */
+            IotMqtt_Assert( pConnectInfo->previousSubscriptionCount > 0 );
+
+            status = _IotMqtt_AddSubscriptions( pNewMqttConnection,
+                                                2,
+                                                pConnectInfo->pPreviousSubscriptions,
+                                                pConnectInfo->previousSubscriptionCount );
         }
     }
 
-    /* Convert the connect info and will info objects to an MQTT CONNECT packet. */
-    status = _getMqttConnectSerializer( pNetworkInfo->pMqttSerializer )( pConnectInfo,
-                                                                         &( pOperation->u.operation.pMqttPacket ),
-                                                                         &( pOperation->u.operation.packetSize ) );
-
-    if( status != IOT_MQTT_SUCCESS )
+    if( status == IOT_MQTT_SUCCESS )
     {
-        goto cleanup;
-    }
+        /* Convert the connect info and will info objects to an MQTT CONNECT packet. */
+        status = _getMqttConnectSerializer( pNetworkInfo->pMqttSerializer )( pConnectInfo,
+                                                                             &( pOperation->u.operation.pMqttPacket ),
+                                                                             &( pOperation->u.operation.packetSize ) );
 
-    /* Check the serialized MQTT packet. */
-    IotMqtt_Assert( pOperation->u.operation.pMqttPacket != NULL );
-    IotMqtt_Assert( pOperation->u.operation.packetSize > 0 );
+        if( status == IOT_MQTT_SUCCESS )
```

Comment:
Having an `if( status == IOT_MQTT_SUCCESS )` within another `if( status == IOT_MQTT_SUCCESS )` looks a bit weird to me. Maybe move it to its own block and reduce nesting complexity at the same time.